torch/utils/_benchmark/utils/timer.py

```diff
@@ -75,7 +76,37 @@ def repeat(self, repeat=-1, number=-1):
     def autorange(self, callback=None):
         raise NotImplementedError("See `Timer.blocked_autorange.`")
 
-    def blocked_autorange(self, callback=None, min_run_time=0.2):
+    def _threaded_measurement_loop(self, hook, min_run_time: float, max_run_time: float, callback=None):
+        # hook should take times [list of float] and return (boolean can_stop_now, time_taken_last_measure)
+        total_time = 0.0
+        can_stop = False
+        times = []
+        with common.set_torch_threads(self._num_threads):
+            while (total_time < min_run_time) or (not can_stop):
+                can_stop, time_spent = hook(times)
+                times.append(time_spent)
+                total_time += time_spent
+                if callback:
+                    callback(number, time_taken)
+                if max_run_time and total_time > max_run_time:
+                    break
+        return times
+
+    def adaptive_autorange(self, threshold=0.1, max_run_time=None, callback=None, min_run_time=0.1):
```

Comment:
`max_run_time` should have some reasonable default (e.g. 10 seconds) so that the default behavior can't lead to infinite loops.

Reply:
done.